core-bundle/src/Resources/contao/controllers/FrontendIndex.php

```diff
@@ -181,9 +181,9 @@ public function renderPage($pageModel)
 		if ($objPage->alias != '')
 		{
 			$language = \Config::get('addLanguageToUrl') ? '[a-z]{2}(-[A-Z]{2})?/' : '';
-			$suffix = \Config::get('urlSuffix') ? preg_quote(\Config::get('urlSuffix'), '#') : '$';
+			$suffix = \Config::get('urlSuffix') ? preg_quote(\Config::get('urlSuffix'), '#') : '';
 
-			if (preg_match('#^' . $language . $objPage->id . '(' . $suffix . '|/)#', \Environment::get('relativeRequest')))
+			if (preg_match('#^' . $language . $objPage->id . '(' . $suffix . '$|/)#', \Environment::get('relativeRequest')))
```

Comment:
```suggestion
			if (preg_match('#^' . $language . $objPage->id . '(' . $suffix . '($|\?)|/)#', \Environment::get('relativeRequest')))
```

If I’m understandig it correctly it should match `de/123.html`, `de/123.html?foo=bar` and `de/123/subpage.html`